consensus/notifications.go

```diff
@@ -85,27 +30,38 @@ func (s *State) notifySubscribers(cc ConsensusChange) {
 	}
 }
 
-// ConsensusSubscribe returns a channel that will receive a ConsensusChange
-// notification each time that the consensus changes (from incoming blocks or
-// invalidated blocks, etc.).
-//
-// TODO: Depricate in favor not sending a whole diff down the channel.
-func (s *State) ConsensusSubscribe() (alert chan ConsensusChange) {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+// OutputDiffsSince returns a set of output diffs representing how the state
+// has changed since block `id`. OutputDiffsSince will flip the `new` value for
+// diffs that got reversed.
+func (s *State) BlocksSince(id BlockID) (removedBlocks, addedBlocks []BlockID, err error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
 
-	alert = make(chan ConsensusChange)
-	s.consensusSubscriptions = append(s.consensusSubscriptions, alert)
-	return
-}
+	node, exists := s.blockMap[id]
+	if !exists {
+		err = errors.New("block is not known to the state or is an orphan")
+		return
+	}
+
+	// Get all the ids from going backwards to the blockchain.
+	reversedNodes := s.backtrackToBlockchain(node)
+	// Eliminate the last node, which is the pivot node, whose diffs are already
+	// known.
+	reversedNodes = reversedNodes[:len(reversedNodes)-1]
+	for _, reversedNode := range reversedNodes {
+		removedBlocks = append(removedBlocks, reversedNode.Block.ID())
+	}
+
+	// Get all the ids going forward from the pivot node.
+	height := reversedNodes[len(reversedNodes)-1].Height
```

Comment:
The pivot node is the node around which the blockchain is changing. So you go 10a -> 9a -> 8 -> 9b -> 10b -> 11b. 8 is the pivot node. reversedNodes is going to be [10a, 9a, 8]. We pop off 8 b/c it doesn't need to be reversed: [10a, 9a]. Then we add diffs for 10a then 9a, and then we add diffs going forward from 8, diffs for [9b, 10b, 11b]